packages/gatsby-image/src/index.js

```diff
@@ -48,18 +48,54 @@ const convertProps = props => {
   return convertedProps
 }
 
+/**
+ * Checks if fluid or fixed are art-direction arrays.
+ *
+ * @param currentData  {array}   The props to check for images.
+ * @return {boolean}
+ */
+const hasArtDirectionSupport = currentData =>
+  currentData &&
+  Array.isArray(currentData) &&
+  currentData.some(image => typeof image.media !== `undefined`)
+
+/**
+ * Tries to detect if a media query matches the current viewport.
+ * @param media   {string}  A media query string.
+ * @return {boolean}
+ */
+const matchesMedia = ({ media }) =>
+  media ? isBrowser && !!window.matchMedia(media).matches : false
+
 /**
  * Find the source of an image to use as a key in the image cache.
  * Use `the first image in either `fixed` or `fluid`
  * @param {{fluid: {src: string}[], fixed: {src: string}[]}} args
  * @return {string}
  */
 const getImageSrcKey = ({ fluid, fixed }) => {
-  const data = (fluid && fluid[0]) || (fixed && fixed[0])
+  const data = fluid ? getCurrentSrcData(fluid) : getCurrentSrcData(fixed)
 
   return data.src
 }
 
+/**
+ * Returns the current src - Preferably with art-direction support.
+ * @param currentData  {array}   The fluid or fixed image array.
+ * @return {*}
```

Comment:
This PR changes the objects to have an optional `media` property so please reflect this in hte JSdocs.

```
 * @param currentData  {{src: string, media?: string}[]}   The fluid or fixed image array.
```

Since it returns an element of this array in all branches, you can replace the star with that type (`{src: string, media?: string}`), too.

(This goes for all the jsdocs here :) )